File: haskell/day1/main.hs
part1 :: [Int] -> Int
part1 xs = length . filter (== True) $ zipWith (<) xs (tail xs)

part2 :: [Int] -> [Int]
part2 xs = zipWith3 (\x y z -> x + y + z) xs (tail xs) (tail $ tail xs)

main :: IO ()
main = interact $ show . part1 . part2 . (map read . lines)

File: rust/day1/main.rs
use std::fs;


fn main () {
  let input: Vec<i32> = fs::read_to_string("input.txt").expect("Unable to read file").split("\n").map(|x| x.parse::<i32>().unwrap()).collect();
  println!("{}", part1(&input));
  println!("{}", part2(&input));
}

fn part1 (input: &Vec<i32>) -> i32 {
  let mut acc = 0;
  for i in 1..input.len() {
    if input[i] > input[i - 1] {
      acc += 1;
    }
  }
  acc
}

fn part2 (input: &Vec<i32>) -> i32 {
  let mut acc = 0;
  for i in 3..input.len() {
    let last3 = input[i] + input[i - 1] + input[i - 2];
    let last4 = input[i - 1] + input[i -2] + input[i - 3];

    if last3 > last4 {
      acc += 1;
    }
  }
  acc
}